### PR TITLE
libobs: Add new signals and private support for output, encoder and service

### PR DESCRIPTION
**Pull request by aet**
### Description

- Add obs_source_create_private equivalent for output, encoder and service and update UI to use it for test instances
- Add obs_source_dosignal equivalent for output, encoder and service
- Add source_create/source_destroy signal equivalent for output, encoder and service
- Add source_update/update signals when obs_output_update called

### Motivation and Context

This is a follow-up modification for https://github.com/obsproject/obs-studio/pull/7797

I realized that while I can use a tick callback to continuously probe when a service/output instance has changed or been created at all, it's not really efficient when we could have a global signals similar to what sources have.

`obs_output_t` has `starting` and `start` signals, but they're both triggered too late to do any last minute encoder/output configuration inspection or modifications, so I added a new signal `update` to handle this. This also allows me to change the main recording/streaming output instances to use a custom view, if needed, to further improve the flexibility of OBS.

### How Has This Been Tested?

- Tested with a custom plugin on a Mac and a PC

### Types of changes

- New feature (non-breaking change which adds functionality)
- Documentation (a change to documentation pages)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Review by RytoEX — CHANGES_REQUESTED** [review]
Please amend the commit message to conform to our 50/72 commit guideline, per our [Commit Guidelines](https://github.com/obsproject/obs-studio/blob/9116ceab6e59ce3dbda9020a8a781a99a38701f1/CONTRIBUTING.rst#commit-guidelines). Please also capitalize the first word after the commit prefix.

**Comment by aet** [comment]
@RytoEX any thoughts on this? There was a request changes toggle earlier but I'm not sure if github lost context where to react when I force pushed the new commit.

Usually various layers of OBS, like source and output follow similar kind of core API conventions, weak references and such, so following the lead to add X create/destroy signals for other than sources should make sense. :)

Not sure if the naming convention for `configure` signal is ok, but it's not a signal that an update just happened, but rather giving you the last chance to react. Currently it's impossible for a regular plugin to affect typical streaming/recording outputs before they're really activated, when starting/start signals are called too late.

**Comment by aet** [comment]
@tt2468 @jp9000 Sorry to ping extra review candidates, but just trying to make sure if there are any design issues with this PR?

**Comment by Fenrirthviti** [comment]
No extra pings are necessary, PRs will be reviewed as we have time. When we have time to review, we will provide comments or updates.

As a note, we've already entered beta cycle for 29, so we're in a feature freeze period while we focus on beta testing and bug fixing there.

**Comment by RytoEX** [comment]
> @RytoEX any thoughts on this? There was a request changes toggle earlier but I'm not sure if github lost context where to react when I force pushed the new commit.
> 
> Usually various layers of OBS, like source and output follow similar kind of core API conventions, weak references and such, so following the lead to add X create/destroy signals for other than sources should make sense. :)
> 
> Not sure if the naming convention for `configure` signal is ok, but it's not a signal that an update just happened, but rather giving you the last chance to react. Currently it's impossible for a regular plugin to affect typical streaming/recording outputs before they're really activated, when starting/start signals are called too late.

I'm not familiar enough with the topic area to give a meaningful review. I defer to others on this.

**Comment by aet** [comment]
Ok thank you, I'll wait patiently and happy to work on feedback if there's any. :) 

The only reason I asked is because my original PR got split into two separate PRs (separate features but helping towards the same goal) and the first part made into 29 beta 1. It's a good start but not quite enough to fully utilize the feature as-is from a standalone plugin point of view. Use of these new signals would make the user experience more reliable than relying on probing for something in the render thread ticks and hoping your timing is good when UI thread decides to start streaming.

**Comment by tytan652** [comment]
> - Add obs_source_dosignal equivalent for output, encoder and service
> - Add source_create/source_destroy signal equivalent for output, encoder and service

This is an API breakage, output, encoder and service<sup>1</sup> have actually a private-like behavior (no global OBS signal) by default.

OBS Studio and maybe plugins rely on this private-like behavior, suddenly making them public is a breakage.

In the settings window and the auto-wizard, no service, encoder and output should be public, maybe even all of them in OBS Studio should not be public.

> - Add configure signal for obs_output_t to allow reconfiguring/settings inspection and validation

Did you consider signals similar to `"source_update"`/`"update"` ?

Because modifying the output and encoder while in code the output is starting, it just feels wrong.

-----
1: Service do have a `create_private` function but still no global signal, so public and private behavior are mostly identical.

**Comment by aet** [comment]
@tytan652 Everyone knows that APIs are painful, when there's too much, when there's too little and how to pace adding new functionality with longterm compatibility in mind. Within current OBS architecture, there's always some corner cases where a plugin X just gets to do special casing all over the code base, and being part of the core OBS source code helps. For many others, they'll need to do a fork with these plugin API limitations.

How can adding new signals be an API breakage if you're required to listen to them and act upon those signals after they get triggered?

The problem for 3rdparty plugins and UI layer of OBS is, that once you've launched OBS, the service/output/encoder related instances are created during first recording/streaming. There's nothing you can listen to. You can poll for their existence through a tick callback, add signal handlers, but at that point it's usually too late already unless you get the info immediately.

obs-output.c has `starting` signal, but it's also too late for plugins to react on anything, if they even know their handle.

`success = output->info.start(output->context.data);` gets called way before `starting` that's the last moment for anyone to apply additional encoder validations/overrides or change the obs_view used through obs_encoder_set_video().

I'm seeing some plugins doing their own start streaming/recording buttons, maintaining encoder settings and catching up what OBS does with the encoder settings, but I think many use cases could be solved by simple signals.

**Comment by tt2468** [comment]
> This is an API breakage, output, encoder and service1 have actually a private-like behavior (no global OBS signal) by default.

I don't see how this is any kind of breaking change, let alone one which I think poses any significant issue. Perhaps you meant that it is a behavioral change? While it is true that there is no public/private for these classes, I don't see how that would imply that they should always be treated as private.

As an example, we have `obs_enum_sources()` which does exclude private sources, but there's also `obs_enum_all_sources()`, which does not discriminate. We also of course have `obs_enum_outputs()` and `obs_enum_encoders()`.

Regarding your last point- sure, but we aren't dealing with UI, we're dealing with an ABI. There's a million and a half ways to break OBS via plugins that we can nitpick all day long, but I don't see how this is realistically anywhere near the top of the totem pole of exposing private parts of OBS.

**Comment by tytan652** [comment]
In case it was not understood, I'm not against adding global signals to output/encoder/service.
But some of them created in the UI or in a first-party, may require to be turned into a private variant to not emit global signals (e.g. Settings, Auto-Wizard…).

**Comment by aet** [comment]
Due to reasons, my plugin in progress has a core variant that has to work without UI/Qt/obs-frontend-api bits but it also builds a UI plugin variant dedicated to all things UI/Qt only when available.

I can work further on this PR by adding private bits to the API and modify some parts under the UI code to use these new private variants.

I was debating earlier if `configure` vs `update` for the signal name is the way to go. `update` is more consistent with the existing things, I'm ok either way

**Comment by aet** [comment]
@tytan652 sorry, updating the PR took a while. As discussed, the new squashed commit adds create_private variants, updates UI code to use them for test_ stuff and renames `configuration` signal to `update`

**Review by tytan652 — COMMENTED** [review]
Thank you for adding `create_private` variants, but you should not put all the change in one commit.

I feel like I'm making you make this PR bigger, because actually I was only sharing concerns.

So for now, split "signals additions" and "private additions" in two separated commits.

For the following, please wait for more feedback before making more changes.

Enum functions (`obs_enum_services`, `obs_enum_outputs`, `obs_enum_encoders`) list private ones because they are the same as `obs_enum_all_sources` and not `obs_enum_sources`.

File: docs/sphinx/reference-outputs.rst
Comment (aet) [reply docs/sphinx/reference-outputs.rst]: Oh now I got it, I somehow totally missed obs_output_update existance. I changed the signaling to do output_update/update similar to sources and it does indeed fill my needs, while overall being less hacky. Thanks!

**Comment by aet** [comment]
No worries, I split the commits, including the small UI tweak to start using the new api. The commits for create/destroy signal and output_update/update signals are also now separate.

**Review by tytan652 — CHANGES_REQUESTED** [review]
- Commit n°1:
  - In consequence of the creation of private variants, `obs_enum` functions that I listed earlier require to enumerate only non-private objects (You also written in the doc that they do that).
    - I don't think `obs_all_enum` variants needs to be created for now so skip them.
  - The commit name needs to be shortened (up to 75 character IIRC) and "update UI code to use it" can be removed since it is no longer the case.

- Commit n°2:
  - You missed `create` encoders and output in `window-basic-auto-config-test.cpp` all of them in this file need to be private.

- Commit n°3 :
  - The commit name needs to be shortened (up to 75 character IIRC)

- Commit n°4:
  - Typo in the commit name "source_update" -> "output_update"

**Comment by aet** [comment]
@tytan652 sorry got lost, I took the obs_enum documentation mention away from the docs since it was accurate, but did you mean tweaking something in obs_enum itself, possibly adding `bool all` parameter, so that obs_enum_all_sources could do other things than the rest, that are all using obs_enum internally? If there ever would be an _all variant, there wouldn't be changed behavior later

**Comment by tytan652** [comment]
> did you mean tweaking something in obs_enum itself, possibly adding `bool all` parameter, so that obs_enum_all_sources could do other things than the rest, that are all using obs_enum internally? If there ever would be an _all variant, there wouldn't be changed behavior later

I mean change `obs_enum_` functions to skip private object like it's done for sources and scenes.

Since outputs, encoders and services didn't had "private variant", their `obs_enum_` functions didn't need to filter them. Now that that those private exist those `obs_enum_` should behave like `obs_enum_sources`/`obs_enum_scenes` by skipping private objects.

`obs_enum()` should not be modified since it meant for filter-less enumeration. And I don't think we need `obs_enum_all_` for them right now.

**Comment by aet** [comment]
> I mean change `obs_enum_` functions to skip private object like it's done for sources and scenes.

Added static helper `obs_enum_public` that's equal to `obs_enum`, except it checks private being false and upgraded obs_enum_ for services, encoders and outputs to use it. Restored documentation back to where it was.

**Comment by tytan652** [comment]
Thank you for making the change. I think all my concerns about the signal additions are addressed.

Still needs testing and review of the code.

**Comment by aet** [comment]
@RytoEX @tytan652 Any ideas who should be nominated to follow with the review? I'm just worried that the PR gets lost, as glancing over the PR list, this is is still saying "changes requested" so maybe it's being skipped over. It looks like once you're force pushing a new revision to github, you should always remember to mark requested changes done before doing so. Can't see anything on my end to mark changes done

**Comment by tytan652** [comment]
> I'm just worried that the PR gets lost,

Don't worry, it will not. It may just take some time.

> this is is still saying "changes requested"

This is because RytoEx did a "Request Changes" (without "sub-conversation") comment about commit name lengths that you fixed recently, so don't worry about that.

----

Aside from that, would you mind changing the PR title to something like (not verbatim):
>  libobs: Add new signals and private for output, encoder and service

To match what the PR is actually adding and avoid confusion.

**Comment by RytoEX** [comment]
> @RytoEX @tytan652 Any ideas who should be nominated to follow with the review? I'm just worried that the PR gets lost, as glancing over the PR list, this is is still saying "changes requested" so maybe it's being skipped over. It looks like once you're force pushing a new revision to github, you should always remember to mark requested changes done before doing so. Can't see anything on my end to mark changes done

Maybe @tt2468 ? This area of code is outside my wheelhouse. I only gave a review on commit message compliance. As @tytan652 said, that GitHub still shows "changes requested" is simply how GitHub works. I asked for changes, and I have not provided a new approving review to override the review state. I'm generally uncomfortable providing approvals for just fixing commit message compliance because it's not easy to see on what basis the approval was granted (just commit message compliance vs. code quality vs. PR overall), and I don't want to confuse those with merge rights with that. In the meantime, I've dismissed my review to clear the "changes requested".

**Pull request closed without merging**
